### PR TITLE
Deployment pod metrics

### PR DESCRIPTION
**Pull request by rahulbandariSensei**
What this PR does / why we need it:

We are not getting important metrics:"privileged","readiness probe","liveness probe" for Pods from Kube-state-metrics. I have checked the pods docs(https://github.com/kubernetes/kube-state-metrics/blob/master/docs/pod-metrics.md)  and found that the above important metrics are missing. Also, We need the metric "kube_deployment_status_replicas_ready", which is not supported as of now as per deployment docs(https://github.com/kubernetes/kube-state-metrics/blob/master/docs/deployment-metrics.md). We need the above information for monitoring the k8s cluster using Kube-state-metrics.

Fixes [kubernetes#1508](https://github.com/kubernetes/kube-state-metrics/issues/1508)

### CODE REVIEW

File: internal/store/builder.go [hunk 1]
```diff
@@ -40,7 +41,6 @@ import (
 	vpaclientset "k8s.io/autoscaler/vertical-pod-autoscaler/pkg/client/clientset/versioned"
 	clientset "k8s.io/client-go/kubernetes"
 	"k8s.io/client-go/tools/cache"
-	"k8s.io/klog/v2"
```
Comment (fpetkovski) [inline internal/store/builder.go:43]: Could you revert this change to keep the imports sorted?

File: internal/store/pod.go [hunk 2]
```diff
@@ -89,9 +89,99 @@ func podMetricFamilies(allowLabelsList []string) []generator.FamilyGenerator {
...
+		createPodReadinessProbeFamilyGenerator(),
+		createPodLivenessProbeFamilyGenerator(),
+		createPodPrivilegedFamilyGenerator(),
 	}
 }
 
+func createPodPrivilegedFamilyGenerator() generator.FamilyGenerator  {
+	return *generator.NewFamilyGenerator(
+		"kube_pod_container_privileged",
+		"Tells whether the container of the pod is in privilege mode.",
```
Comment (fpetkovski) [inline internal/store/pod.go:101]: I order to be consistent with other description messages, I would replace tells with describes:
```
Describes whether the containers readiness check succeeded.
```

Same for the other added metrics.

File: docs/deployment-metrics.md [hunk 3]
```diff
@@ -3,6 +3,7 @@
 | Metric name| Metric type | Labels/tags | Status |
 | ---------- | ----------- | ----------- | ----------- |
 | kube_deployment_status_replicas | Gauge | `deployment`=&lt;deployment-name&gt; <br> `namespace`=&lt;deployment-namespace&gt; | STABLE |
+| kube_deployment_status_replicas_ready | Gauge | `deployment`=&lt;deployment-name&gt; <br> `namespace`=&lt;deployment-namespace&gt; | STABLE |
```
Comment (fpetkovski) [inline docs/deployment-metrics.md:6]: Could you also add the other metrics to the documentation?

**Review by fpetkovski — CHANGES_REQUESTED** [review]
Thanks for the contribution, this looks quite good already. Besides the needed rebase, I also left a few smaller suggestions.

File: docs/deployment-metrics.md (hunk 3 above)
Comment (rahulbandariSensei) [reply docs/deployment-metrics.md]: Done.

File: internal/store/pod.go (hunk 2 above)
Comment (rahulbandariSensei) [reply internal/store/pod.go]: done.

File: internal/store/builder.go (hunk 1 above)
Comment (rahulbandariSensei) [reply internal/store/builder.go]: done.

**Comment by rahulbandariSensei** [comment]
@fpetkovski  @mrueg @dgrisonnet Can you please review this PR. TIA.

**Comment by fpetkovski** [comment]
Looks like some of the tests are failing with the latest changes

File: internal/store/pod.go [hunk 4]
```diff
@@ -90,9 +90,99 @@ func podMetricFamilies(allowAnnotationsList, allowLabelsList []string) []generat
...
+				Metrics: ms,
+			}
+		}),
+	)
+}
+
+func createPodLivenessProbeFamilyGenerator() generator.FamilyGenerator {
+	return *generator.NewFamilyGenerator(
+		"kube_pod_container_liveness_probe",
+		"Describes whether the containers liveness check succeeded.",
```
Comment (dgrisonnet) [inline internal/store/pod.go:131]: According to your implementation, this isn't what is exposed. `Pod.Spec.Containers.LivenessProbe` will tell you whether a liveness probe is set up or not for this container.

File: internal/store/pod.go [hunk 5]
```diff
@@ -90,9 +90,99 @@ func podMetricFamilies(allowAnnotationsList, allowLabelsList []string) []generat
...
+		metric.Gauge,
+		"",
+		wrapPodFunc(func(p *v1.Pod) *metric.Family {
+			ms := make([]*metric.Metric, len(p.Spec.Containers))
+			labelKeys := []string{"container"}
+
+			for i, c := range p.Spec.Containers {
+				livenessProbe := 0
```
Comment (dgrisonnet) [inline internal/store/pod.go:139]: this should be a boolean

File: internal/store/pod.go [hunk 6]
```diff
@@ -90,9 +90,99 @@ func podMetricFamilies(allowAnnotationsList, allowLabelsList []string) []generat
 		createPodStatusScheduledFamilyGenerator(),
 		createPodStatusScheduledTimeFamilyGenerator(),
 		createPodStatusUnschedulableFamilyGenerator(),
+		createPodReadinessProbeFamilyGenerator(),
+		createPodLivenessProbeFamilyGenerator(),
+		createPodPrivilegedFamilyGenerator(),
```
Comment (dgrisonnet) [inline internal/store/pod.go:95]: could you sort these? It would be great to also sort the definition to keep this file easier to navigate.

**Comment by mindw** [comment]
Hi, just a quick question, how useful are the probe metrics given kubelet exposes them already as counters on  `/metrics/probes/` ? 
These seem like duplicates and far less realiable?
```
k get --raw='/api/v1/nodes/<node name>:10250/proxy/metrics/probes'
...
prober_probe_total{container="jaeger-agent",namespace="jaeger",pod="jaeger-agent-9dh5k",pod_uid="1a2488dc-6d3b-4cd1-b282-0c53ec76c315",probe_type="Liveness",result="successful"} 118243
prober_probe_total{container="jaeger-agent",namespace="jaeger",pod="jaeger-agent-9dh5k",pod_uid="1a2488dc-6d3b-4cd1-b282-0c53ec76c315",probe_type="Readiness",result="successful"} 118244
prober_probe_total{container="jaeger-collector",namespace="jaeger",pod="jaeger-collector-677bfd8977-rczvf",pod_uid="51c7c6ab-b0cd-4b6e-8ce1-dc0ec5d06f90",probe_type="Liveness",result="successful"} 118243
prober_probe_total{container="jaeger-collector",namespace="jaeger",pod="jaeger-collector-677bfd8977-rczvf",pod_uid="51c7c6ab-b0cd-4b6e-8ce1-dc0ec5d06f90",probe_type="Readiness",result="successful"} 118243
prober_probe_total{container="kafka-broker",namespace="default",pod="connect-kafka-1",pod_uid="f55ebf8d-f173-44a4-ab39-92ba0dbf39c2",probe_type="Liveness",result="successful"} 240006
prober_probe_total{container="kafka-broker",namespace="default",pod="connect-kafka-1",pod_uid="f55ebf8d-f173-44a4-ab39-92ba0dbf39c2",probe_type="Readiness",result="successful"} 240008
prober_probe_total{container="kafka-broker",namespace="default",pod="notification-kafka-0",pod_uid="55ff0e20-9228-4901-ad80-64125e5aa0ff",probe_type="Liveness",result="successful"} 238785
prober_probe_total{container="kafka-broker",namespace="default",pod="notification-kafka-0",pod_uid="55ff0e20-9228-4901-ad80-64125e5aa0ff",probe_type="Readiness",result="successful"} 238787
...
```

**Comment by mrueg** [comment]
@brancz had made a suggestion here: https://github.com/kubernetes/kube-state-metrics/issues/1508#issuecomment-879026300 can you incorporate that into your PR?

**Comment by dgrisonnet** [comment]
@rahulbandariSensei based on @mindw comment https://github.com/kubernetes/kube-state-metrics/pull/1552#issuecomment-922840572, do you still have a use-case for the metrics that you are introducing? If not I would be in favor of closing this PR since the metrics already exist.

**Comment by chicofranchico** [comment]
Does the prober_probe_total include the privileged label? If not, would make sense to limit this PR to that specifically. Thoughts?
